sdk/tests/core/test_az_hfsm_pipeline.c

```diff
@@ -0,0 +1,376 @@
+
+    case P_INBOUND_0:
+      pinbound0++;
+      break;
+
+    case S_OUTBOUND_0:
+      soutbound0++;
+      ret = az_hfsm_pipeline_send_outbound_event((az_hfsm_policy*)me, soutbound1_evt);
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_hfsm_state_handler p01_get_parent_test(az_hfsm_state_handler child_state)
+{
+  az_hfsm_state_handler parent_state;
+
+  if ((child_state == P01Root))
+  {
+    parent_state = NULL;
+  } else 
+  {
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static az_hfsm_state_handler p02_get_parent_test(az_hfsm_state_handler child_state)
+{
+  az_hfsm_state_handler parent_state;
+
+  if ((child_state == P02Root))
+  {
+    parent_state = NULL;
+  } else 
+  {
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static az_hfsm_state_handler p03_get_parent_test(az_hfsm_state_handler child_state)
+{
+  az_hfsm_state_handler parent_state;
+
+  if ((child_state == P03Root))
+  {
+    parent_state = NULL;
+  } else 
+  {
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static void test_az_hfsm_pipeline_init(void** state)
+{
+  (void)state;
+
+  // Init P01 policy
+  assert_int_equal(az_hfsm_init((az_hfsm*) &P01, P01Root, p01_get_parent_test), AZ_OK);
+  assert_true(P01.hfsm._internal.current_state == P01Root);
+  assert_true(refp01root == 1);
+
+  // Init P02 policy
+  assert_int_equal(az_hfsm_init((az_hfsm*) &P02, P02Root, p02_get_parent_test), AZ_OK);
+  assert_true(P02.hfsm._internal.current_state == P02Root);
+  assert_true(refp02root == 1);
+
+  // Init P03 policy
+  assert_int_equal(az_hfsm_init((az_hfsm*) &P03, P03Root, p03_get_parent_test), AZ_OK);
+  assert_true(P03.hfsm._internal.current_state == P03Root);
+  assert_true(refp03root == 1);
+
+  P01.inbound = &P02;
+  P02.outbound = &P01;
+  P02.inbound = &P03;
+  P03.outbound = &P02;
+
+  assert_int_equal(az_hfsm_pipeline_init(&az_hfsm_pipeline_test, &P01, &P03), AZ_OK);
+}
+
+static void test_az_hfsm_pipeline_post_outbound(void** state)
+{
+  (void)state;
+
+  assert_int_equal(az_hfsm_pipeline_post_outbound_event(&az_hfsm_pipeline_test, poutbound0_evt),
+   AZ_OK);
+  assert_true(poutbound0 == 1);
+}
+
+static void test_az_hfsm_pipeline_post_inbound(void** state)
+{
+  (void)state;
+
+  assert_int_equal(az_hfsm_pipeline_post_inbound_event(&az_hfsm_pipeline_test, pinbound0_evt),
+   AZ_OK);
+  assert_true(pinbound0 == 1);
+}
+
+static void test_az_hfsm_pipeline_send_inbound(void** state)
+{
+  (void)state;
+  
+  assert_int_equal(az_hfsm_pipeline_post_outbound_event(&az_hfsm_pipeline_test, sinbound0_evt),
+   AZ_OK);
+  assert_true(sinbound0 == 1);
+  assert_true(sinbound1 == 1);
+}
+
+static void test_az_hfsm_pipeline_send_inbound_failure(void** state)
+{
+  (void)state;
+
+  assert_int_equal(az_hfsm_pipeline_post_outbound_event(&az_hfsm_pipeline_test, sinbound2_evt),
+   AZ_OK);
+  assert_true(sinbound2 == 1);
+  assert_true(sinbound3 == 2);
```

Comment:
are these all purposely not reset between tests? Should they be, so that the tests aren't depended on running them in series?

Reply:
Added the reset for non-ref variables